### PR TITLE
[bitnami/pgpool] extend health check configuration

### LINKED ISSUE #12263: [bitnami/postgresql-ha] Postgresql-HA ends up with multiple primaries on node changes
### Name and Version

bitnami/postgresql-ha

### What steps will reproduce the bug?

Extremely hard to replicate, but we've seen it often enough now.

Essentially, the root cause is we change our nodes weekly in our cluster. We add one new node, uncordon + drain an old node, then add the new one via Cluster API.

Occasionally, about once a month, one of our databases ends up in a state where 2 postgresql pods from the postgresql-ha chart believe they are primary. This is usually resolves by just picking a pod and restarting it, then the postgresql pods will then re-nominate the primary.

### Are you using any custom parameters or values?

_No response_

### What is the expected behavior?

Would expect postgresql-ha to not get in this state. Totally understand there is a customXProbe value exposed, but was hoping we could land on a better default.

### What do you see instead?

I believe altering the liveness probe or adding a startup to ensure that *-postgresql-0 is primary and *-postgresql-X is secondary would be best.

I believe this is related and could fix #8062 as well.

Could of ways to implement this:
1) psql -U postgres -c "SELECT pg_is_in_recovery()" is true for the primary and false for all others
2) psql -U postgres -c "SELECT pg_stat_receiver" returns contents if primary and there are secondaries. "SELECT pg_stat_wal_receiver" would return content on all secondaries.

For context, we always check via method 2, next time it occurs I'll be sure to check via 1 to ensure both are good ways to verify.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by pcerny**
### Description of the change

Introduce the following changes:
- Introduce PGPOOL_CONNECT_TIMEOUT variable and set default to 10000ms.
  User can setup pgpool to fail faster when trying to connect to unreachable
  PG node.
- Introduce PGPOOL_HEALTH_CHECK_PSQL_TIMEOUT variable with default to 15.
  User can control "show pool_nodes" command timeout in healthcheck
  function.
- Only nodes in pgpool state DOWN which are UP in reality are attached.

### Benefits

Makes pgpool pod configuration more versatile. It is used by PR https://github.com/bitnami/charts/pull/14175

### Possible drawbacks

Do not know.

### Applicable issues

It fixes https://github.com/bitnami/charts/issues/12263.

### Additional information

N/A

### CODE REVIEW

File: bitnami/pgpool/4/debian-11/rootfs/opt/bitnami/scripts/libpgpool.sh [hunk 1]
```diff
@@ -71,10 +71,11 @@ export PGPOOL_ENABLE_STATEMENT_LOAD_BALANCING="${PGPOOL_ENABLE_STATEMENT_LOAD_BA
...
 export PGPOOL_HEALTH_CHECK_USER="${PGPOOL_HEALTH_CHECK_USER:-$PGPOOL_SR_CHECK_USER}"
-export PGPOOL_HEALTH_CHECK_PERIOD="${PGPOOL_HEALTH_CHECK_PERIOD:-30}"
-export PGPOOL_HEALTH_CHECK_TIMEOUT="${PGPOOL_HEALTH_CHECK_TIMEOUT:-10}"
-export PGPOOL_HEALTH_CHECK_MAX_RETRIES="${PGPOOL_HEALTH_CHECK_MAX_RETRIES:-5}"
-export PGPOOL_HEALTH_CHECK_RETRY_DELAY="${PGPOOL_HEALTH_CHECK_RETRY_DELAY:-5}"
+export PGPOOL_HEALTH_CHECK_PERIOD="${PGPOOL_HEALTH_CHECK_PERIOD:-13}"
+export PGPOOL_HEALTH_CHECK_TIMEOUT="${PGPOOL_HEALTH_CHECK_TIMEOUT:-2}"
+export PGPOOL_HEALTH_CHECK_MAX_RETRIES="${PGPOOL_HEALTH_CHECK_MAX_RETRIES:-3}"
+export PGPOOL_HEALTH_CHECK_RETRY_DELAY="${PGPOOL_HEALTH_CHECK_RETRY_DELAY:-2}"
+export PGPOOL_CONNECT_TIMEOUT="${PGPOOL_CONNECT_TIMEOUT:-2000}"
```
Comment (fmulero) [inline bitnami/pgpool/4/debian-11/rootfs/opt/bitnami/scripts/libpgpool.sh:78]: I've been taking a look to the configuration and I think 2s is too agressive for the `PGPOOL_CONNECT_TIMEOUT` and the `PGPOOL_HEALTH_CHECK_RETRY_DELAY` should be greater than `PGPOOL_CONNECT_TIMEOUT` to handle properly errors triggered by that threshold.

What do you think about this possible default config?:

```suggestion
export PGPOOL_HEALTH_CHECK_PERIOD="${PGPOOL_HEALTH_CHECK_PERIOD:-13}"
export PGPOOL_HEALTH_CHECK_TIMEOUT="${PGPOOL_HEALTH_CHECK_TIMEOUT:-5}"
export PGPOOL_HEALTH_CHECK_MAX_RETRIES="${PGPOOL_HEALTH_CHECK_MAX_RETRIES:-2}"
export PGPOOL_HEALTH_CHECK_RETRY_DELAY="${PGPOOL_HEALTH_CHECK_RETRY_DELAY:-2}"
export PGPOOL_CONNECT_TIMEOUT="${PGPOOL_CONNECT_TIMEOUT:-3000}"
```

**Review by fmulero — CHANGES_REQUESTED** [review]
Hi @pcerny Thanks a lot for the contribution. It looks a great improvement.

It LGTM but I've left a comment about the default config. Also could you update your branch with the latest changes from the `main` branch? The container is not spinning up correctly from your branch and I think that problem  is already fixed in the main branch.

File: bitnami/pgpool/4/debian-11/rootfs/opt/bitnami/scripts/libpgpool.sh (hunk 1 above)
Comment (rafariossaa) [reply bitnami/pgpool/4/debian-11/rootfs/opt/bitnami/scripts/libpgpool.sh:78]: Hi,
  I agree with @fmulero, I wouldn't go so aggressive by default and leave them as they are. As they are configurable by the user, they could be set easily to other values by the user if they need it.

**Comment by pcerny** [comment]
TL;DR I would like to rework this PR because I have uncovered race condition in current solution. New solution will be based on marking PG headless service with flag `publishNotReadyAddresses: true`. If you have any thought or experience with that, please share them.

By yet another testing of this PR, I have uncovered that liveness check update contains race condition. It is working only under specific conditions. Most of my tests satisfied those condition so I did not uncover it earlier.
Below is simple rolling upgrade scenario for 3 PG backends which results in liveness check failure and by that to restart of pgpool:
- Starting rolling upgrade in state PG-0 UP (primary), PG-1 UP (standby), PG-2 UP (standby)
- Update PG-2, pgpoolu marks PG-2 DOWN
- As soon as new instance of PG-2 is ready, PG-1 is being terminated and pgpool can connect to PG-2. After some time, pgpool 's liveness probe attaches PG-2 and PG-2 is marked UP (standby)
- During PG-1 termination pgpool recognizes it as DOWN
- As soon as new instance of PG-1 is ready, PG-0 is terminated. PG-0 is primary. pgpool is in state PG-2 UP (standby), PG-1 DOWN (standby), PG-0 UP (primary)
- Repmgr detects PG-0 is unreachable and switches primary to PG-1. In-parallel, pgpool marks PG-0 as DOWN. pgpool is in state PG-2 UP (standby), PG-1 DOWN (standby), PG-0 DOWN (primary)
- pgpool refuses pcp interface actions attach_node or promote_node generally because it is in search of primary in UP state (DOWN state nodes are not included into search)
- pgpool's liveness probe is failing, pgpool is eventually restarted

Problem is that pgpool does not mark PG-1 as UP before PG-0 is marked as DOWN. This two operations are executed in parallel. In order to serialize them, one needs to delay termination of PG-0 which is always triggered by PG-1 achieving ready state. Similarly, pgpool can reach PG-1 once it gets ready because services showing only ready PODs (by default).

Luckily for me, our k8s guru advised me, there exists k8s service option `publishNotReadyAddresses: true`. I did various tests with this flag enabled. All of them successful. I can share more details from tests to answer questions like how repmgr works when PG headless service has this flag set, ideally in form of article. But I am fighting to find a free time slot for that.

 I will try to update both PRs in 2 days. If you have any comments sooner, please let me know.

**Comment by fmulero** [comment]
Thanks a lot @pcerny! I am eager to see your next steps

`publishNotReadyAddresses: true`is very useful to discover new members in cluster like this case. You have to be careful if you are exposing that service to your clients because you could reach a non ready pod.

**Comment by pcerny** [comment]
I decided to keep defaults to minimize impact on current PG-HA setups. So this PR adds just two new env variables as described in the first PR comment.

Related PR https://github.com/bitnami/charts/pull/14175 builds on it and introduce ability to setup PG headless service with `publishNotReadyAddresses: true`. I leave comment about how I test it in the related PR.

File: bitnami/pgpool/4/debian-11/rootfs/opt/bitnami/scripts/libpgpool.sh [hunk 2]
```diff
@@ -292,21 +294,29 @@ pgpool_attach_node() {
...
+    if [[ "$backends" != "command failed" ]]; then
+        # look up backends that are marked offline and being up - attach only status=down and pg_status=up
+        # situation down|down means PG is not yet ready to be attached
+        for node in $(echo "${backends}" | grep "down|up" | tr -d ' '); do
             IFS="|" read -ra node_info <<< "$node"
             local node_id="${node_info[0]}"
             local node_host="${node_info[1]}"
             if [[ $(PGCONNECT_TIMEOUT=3 PGPASSWORD="${PGPOOL_POSTGRES_PASSWORD}" psql -U "${PGPOOL_POSTGRES_USERNAME}" \
-                -d postgres -h "${node_host}" -p "${PGPOOL_PORT_NUMBER}" -tA -c "SELECT 1" || true) == 1 ]]; then
+                 -d postgres -h "${node_host}" -p "${PGPOOL_PORT_NUMBER}" -tA -c "SELECT 1" || true) == 1 ]]; then
```
Comment (rafariossaa) [inline bitnami/pgpool/4/debian-11/rootfs/opt/bitnami/scripts/libpgpool.sh:313]: Remove this unneeded white space.

File: bitnami/pgpool/README.md [hunk 3]
```diff
@@ -267,10 +267,12 @@ Pgpool configuration:
...
-- `PGPOOL_HEALTH_CHECK_TIMEOUT`: Specifies the timeout in seconds to give up connecting to the backend PostgreSQL if the TCP connect does not succeed within this time. Defaults to `10`.
-- `PGPOOL_HEALTH_CHECK_MAX_RETRIES`: Specifies the maximum number of retries to do before giving up and initiating failover when health check fails. Defaults to `5`.
-- `PGPOOL_HEALTH_CHECK_RETRY_DELAY`: Specifies the amount of time in seconds to sleep between failed health check retries. Defaults to `5`.
+- `PGPOOL_HEALTH_CHECK_PERIOD`: Specifies the interval between the health checks in seconds. Defaults to `13`.
+- `PGPOOL_HEALTH_CHECK_TIMEOUT`: Specifies the timeout in seconds to give up connecting to the backend PostgreSQL if the TCP connect does not succeed within this time. Defaults to `2`.
+- `PGPOOL_HEALTH_CHECK_MAX_RETRIES`: Specifies the maximum number of retries to do before giving up and initiating failover when health check fails. Defaults to `3`.
+- `PGPOOL_HEALTH_CHECK_RETRY_DELAY`: Specifies the amount of time in seconds to sleep between failed health check retries. Defaults to `2`.
```
Comment (rafariossaa) [inline bitnami/pgpool/README.md:273]: I think for this we should keep the previous default values, and set then to other values, in the chart file.
This could break the current behavior for the current users of the image.

File: bitnami/pgpool/README.md [hunk 4]
```diff
@@ -267,10 +267,12 @@ Pgpool configuration:
...
+- `PGPOOL_CONNECT_TIMEOUT`: Specifies the amount of time in milliseconds before giving up connecting to backend using `connect()` system call. Default is `10000`.
+- `PGPOOL_HEALTH_CHECK_PSQL_TIMEOUT`: Defines PGCONNECT_TIMEOUT value for "show pool_nodes" command in "pgpool_healthcheck" function. Default is `15`.
```
Comment (rafariossaa) [inline bitnami/pgpool/README.md:275]: Indicate the time unit (seconds).

**Review by rafariossaa — DISMISSED** [review]

File: bitnami/pgpool/4/debian-11/rootfs/opt/bitnami/scripts/libpgpool.sh (hunk 2 above)
Comment (pcerny) [reply bitnami/pgpool/4/debian-11/rootfs/opt/bitnami/scripts/libpgpool.sh]: Fixed

File: bitnami/pgpool/README.md (hunk 3 above)
Comment (pcerny) [reply bitnami/pgpool/README.md]: Fixed

File: bitnami/pgpool/README.md (hunk 4 above)
Comment (pcerny) [reply bitnami/pgpool/README.md]: Fixed

**Review by fmulero — APPROVED** [review]
LGTM

**Review by fmulero — APPROVED** [review]
